primitives/primproc/batchprimitiveprocessor.cpp

```diff
@@ -541,31 +589,171 @@ void BatchPrimitiveProcessor::addToJoiner(ByteStream& bs)
+                    tlLargeKey.deserialize(bs, storedKeyAllocator);
                     bs >> tlIndex;
-                    tlJoiners[joinerNum]->insert(pair<TypelessData, uint32_t>(tlLargeKey,
-                                                 tlIndex));
+                    bucket = bucketPicker((char *) tlLargeKey.data, tlLargeKey.len, bpSeed) & ptMask;
+                    tmpBuckets[bucket].push_back(make_pair(tlLargeKey, tlIndex));
                 }
                 else
-                    tJoinerSizes[joinerNum]--;
+                    --tJoinerSize;
+            }
+
+            bool done = false, didSomeWork;
+            //uint loopCounter = 0, noWorkCounter = 0;
+            // this loop moves the elements from each vector into its corresponding hash table.
+            while (!done)
+            {
+                //++loopCounter;
+                done = true;
+                didSomeWork = false;
+                for (i = 0; i < processorThreads; ++i)
+                {
+                    if (!tmpBuckets[i].empty())
+                    {
+                        bool gotIt = addToJoinerLocks[joinerNum][i].try_lock();
+                        if (!gotIt)
+                        {
+                            done = false;   // didn't get it, don't block, try the next bucket
+                            continue;
+                        }
+                        for (auto &element : tmpBuckets[i])
+                            tlJoiners[joinerNum][i]->insert(element);
+                        addToJoinerLocks[joinerNum][i].unlock();
+                        tmpBuckets[i].clear();
+                        didSomeWork = true;
+                    }
+                }
+                // if this iteration did no useful work, everything we need is locked; wait briefly
+                // and try again.
+                if (!done && !didSomeWork)
+                {
+                    ::usleep(500 * processorThreads);
+                    //++noWorkCounter;
+                }
+            }
+            //cout << "TL join insert.  Took " << loopCounter << " loops" << endl;
+
+        }
+        else
+        {
+            boost::shared_array<boost::shared_ptr<TJoiner> > tJoiner = tJoiners[joinerNum];
+            uint64_t nullValue = joinNullValues[joinerNum];
+            bool &l_doMatchNulls = doMatchNulls[joinerNum];
+            joblist::JoinType joinType = joinTypes[joinerNum];
+            vector<pair<uint64_t, uint32_t> > tmpBuckets[processorThreads];
+
+            if (joinType & MATCHNULLS)
+            {
+                // this first loop hashes incoming values into vectors that parallel the hash tables.
+                for (i = 0; i < count; ++i)
+                {
+                    /* A minor optimization: the matchnull logic should only be used with
+                     * the jointype specifies it and there's a null value in the small side */
```

Comment:
What do you think if we unwind the loop here to see if compiler switches to SIMD version of cmp instruction to compare 2 or 4 values at a time?

Reply:
It definitely won't as long as we still have -fno-tree-vectorize in our cflags.

Would def be interesting to experiment with things like that, but lets get the functionality in before tinkering.